go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,204 @@
+package ovn
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"reflect"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+)
+
+type EgressDNS struct {
+	// Protects pdMap/namespaces operations
+	lock sync.Mutex
+	// holds DNS entries globally
+	dns *util.DNS
+	// this map holds dnsNames to the dnsEntries
+	dnsEntries map[string]*dnsEntry
+	// allows for the creation of addresssets
+	addressSetFactory AddressSetFactory
+
+	// Report change when Add operation is done
+	added          chan bool
+	stopChan       chan struct{}
+	controllerStop <-chan struct{}
+}
+
+type dnsEntry struct {
+	//this map holds all the namespaces that a dnsName appears in
+	namespaces map[string]struct{}
+	// the current IP addresses the dnsName resolves to
+	dnsResolves []net.IP
+	// the addressSet that contains the current IPs
+	dnsAddressSet AddressSet
+}
+
+func (oc *Controller) NewEgressDNS(controllerStop <-chan struct{}) (*EgressDNS, error) {
+	dnsInfo, err := util.NewDNS("/etc/resolv.conf")
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil, err
+	}
+
+	egressDNS := &EgressDNS{
+		dns:               dnsInfo,
+		dnsEntries:        make(map[string]*dnsEntry),
+		addressSetFactory: oc.addressSetFactory,
+
+		added:          make(chan bool),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	go utilwait.Until(egressDNS.Sync, 0, egressDNS.stopChan)
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(namespace, dnsName string) (AddressSet, error) {
```

Comment:
TODO: dual-stack

Reply:
unless I am mistaken, I do support dual-stack. At this add we just need the namespace and the dnsName the functions in the dns package (namly getIPsAndMinTTL) will get both A and AAAA records and add them to the address sets as available.

Reply:
chatted on slack, the addressSet code deals with dual stack for me